### PR TITLE
incr.comp.: Implement compiler diagnostic persistence.

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR implements storing and loading diagnostics that the compiler generates and thus allows for emitting warnings during incremental compilation without actually re-evaluating the thing the warning originally came from. It also lays some groundwork for storing and loading type information and MIR in the incr. comp. cache.

~~It is still work in progress:~~
- ~~There's still some documentation to be added.~~
- ~~The way anonymous queries are handled might lead to duplicated emissions of warnings. Not sure if there is a better way or how frequent such duplication would be in practice.~~

Diagnostic message duplication is addressed separately in #45519.

r? @nikomatsakis

### CODE REVIEW

File: src/librustc/ty/context.rs [hunk 1]
```diff
@@ -853,6 +853,11 @@ pub struct GlobalCtxt<'tcx> {
 
     pub dep_graph: DepGraph,
 
+    /// This provides access to the incr. comp. on-disk cache for query results.
+    /// Do not access this directly. It is only meant to be used by
+    /// `DepGraph::try_mark_green()` and the query infrastructure in `ty::maps`.
+    pub(crate) on_disk_query_result_cache: maps::OnDiskCache<'tcx>,
```
Comment (nikomatsakis) [inline src/librustc/ty/context.rs]: as an aside, it feels to me like "`pub(crate)` visibility with scary comment" is a sign that this code is not "well aligned" in terms of module structure. We should review at some point.

File: src/test/incremental/warnings-reemitted.rs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// revisions: cfail1 cfail2 cfail3
+// compile-flags: -Coverflow-checks=on
+// must-compile-successfully
```
Comment (nikomatsakis) [inline src/test/incremental/warnings-reemitted.rs]: this test seems good, but I wish we had some way to guarantee that the warning was coming from the red-green code. e.g., some way to assert that the relevant query that emits the warning normally was not running. But... good enough I guess.

**Review by nikomatsakis — APPROVED** [review]

**Comment by kennytm** [comment]
MIPS TLS problem same as that of https://github.com/rust-lang/rust/pull/45187#issuecomment-340216465 and https://github.com/rust-lang/rust/pull/45529#issuecomment-339903319. Legit.

**Comment by alexcrichton** [comment]
The linkage error here I hope will be unblocked with https://github.com/rust-lang/rust/pull/45655

**Pull request merged**